modules/standard/Socket.md

```diff
@@ -0,0 +1,235 @@
+
+
+`ipAddr` instance created will then have getters for all the three arguments i.e. `host`, `port` and `family` which dont take up any space inside the record and are fetched from [sockaddr_storage](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_socket.h.html) field stored in `ipAddr`.
+
+## Connecting to remote sockets
+
+`connect` function will create `tcpConn` instance which will take in the following parameters
+
+**Method:**
+```python
+proc connect(in address:ipAddr, in timeout = new timeval(-1,0)) : tcpConn
+```
+
+**Parameters:**
+
+```
+address: ipAddr - address information of remote socket
+timeout: timeval - defines the time in `(seconds, microseconds)` function should wait before
+                  throwing `TimeoutError`. `(-1, 0)` implies indefinite blocking.
+```
+
+A higher level construct for socket will take in `host` as `string` which can perform DNS resolution and convert to `ipAddr` internally.
+
+**Method:**
+```python
+proc connect(host:string, port, family:IPFamily = IPFamily.IPv4, in timeout = new timeval(-1,0)) : tcpConn
+```
+
+**Parameters:**
+```
+host: String - host name either in dot-dash or domain form default "localhost"
+port: Integer - port number
+family: IPFamily - Internet Protocol version which can be IPFamily.IPv4 or IPFamily.IPv6
+timeout: timeval - defines the time in `(seconds, microseconds)` function should wait before
+                  throwing `TimeoutError`. `(-1, 0)` implies indefinite blocking.
+```
+
+The purpose of family is important for address resolution and converting addr string to standard notation as well as validation. How the family can affect address resolution depends on [getaddrinfo](https://pubs.opengroup.org/onlinepubs/9699919799/functions/getaddrinfo.html) function.
+
+The `tcpConn` instance is basically a chapel `file` instance with added methods aligned to its nature as a socket. All the usual `file` methods are available on it:
+- `reader`
+  will return a reader channel for associated socket
+  ```python
+  proc tcpConn.writer() : channel
+  ```
+- `writer`
+  will return a writer channel for associated socket
+  ```python
+  proc tcpConn.writer() : channel
+  ```
+- `close` calling this on `tcpConn` will close the socket connection and associated `file` too.
+  ```python
+  proc tcpConn.close()
+  ```
+- `addr` this function will return a `ipAddr` object which will contain the `host` and `port` where socket is bound.
+
+## Creating a Listener
+
+`listen` function will create a `tcpListener` Instance which will take in following parameters
+
+**Method:**
+```python
+proc listen(in address:ipAddr, reuseAddr = true, backlog = 5) : tcpListener
+```
+
+**Parameters**
+```
+address: ipAddr - contains family info of socket to create and where to bind it.
+reuseaddr: Boolean - optional parameter, default value = true
+backlog: int - optional parameter, default value = 0
```

Comment:
This conflicts with the declaration above.  Will the default value be 5 or 0?

Reply:
I didn't check on it much but now that I see the value will be will use the `DEFAULT_BACKLOG` as the default which will be:
```chapel
param const DEFAULT_BACKLOG = if SOMAXCONN <= 128 then SOMAXCONN else 128;
```

It's based on how python does it https://bugs.python.org/issue21455 but there was discussion on increasing the backlog which never seemed to gather consensus https://bugs.python.org/issue38699